### PR TITLE
Vrf evaluation commands

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR
* moves the VRF evaluation logic out of consensus and into a separate library
* exposes a GraphQL endpoint for committing to and validating VRF evaluations
* adds commands under `advanced vrf` for generating and verifying VRF commitments
  - these commands operate on a local private key file (where applicable) rather than using GraphQL. Performing many VRF evaluations over GraphQL may significantly harm performance, so this seeks to avoid it
* exposes the `index` of an account in the ledger to GraphQL
  - this is used as one of the inputs to the VRF, so we need to make it available (cc @bkase is this account -> index map something that we need to begin tracking in the archive db?)

Tested the GraphQL commands locally on a test node.

Sample output from running the commands (input is interleaved with output):
```
$ _build/default/src/app/cli/src/mina.exe advanced vrf --help
Commands for vrf evaluations

  mina.exe advanced vrf SUBCOMMAND

=== subcommands ===

  batch-check-witness     Check a batch of vrf evaluation witnesses read on
                          stdin
  batch-generate-witness  Generate a batch of vrf evaluation witnesses from
                          {"globalSlot": _, "epochSeed": _, "delegatorIndex": _}
                          JSON message objects read on stdin
  generate-witness        Generate a vrf evaluation witness
  help                    explain a given subcommand (perhaps recursively)

unknown subcommand --help
$ _build/default/src/app/cli/src/mina.exe advanced vrf generate-witness -help
Generate a vrf evaluation witness

  mina.exe advanced vrf generate-witness

=== flags ===

  --delegator-index NUM            The index of the delegating account in the
                                   epoch ledger
  --epoch-seed SEED                Epoch seed to evaluate the VRF with
  --global-slot NUM                Global slot to evaluate the VRF for
  --privkey-path FILE              File to write private key into (public key
                                   will be FILE.pub)
                                   (alias: -privkey-path)
  [--delegated-stake AMOUNT]       The balance of the delegating account in the
                                   epoch ledger
  [--generate-outputs true|false]  Whether to generate the vrf in addition to
                                   the witness (default: false)
  [--total-stake AMOUNT]           The total balance of all accounts in the
                                   epoch ledger
  [-help]                          print this help text and exit
                                   (alias: -?)

$ _build/default/src/app/cli/src/mina.exe advanced vrf generate-witness --delegator-index 0 --epoch-seed 2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA --global-slot 0 --privkey-path ~/.mina-config/wallets/store/B62qiqVsJm
i5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh
Enter password:
Again to confirm:
{
  "message": {
    "globalSlot": "0",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  },
  "publicKey": "B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh",
  "c":
    "15310236459953163824916066791323397157230465705183945844894713032240475400215",
  "s":
    "24129570863947632229302869722696095657099766803044721376181110670716079373487",
  "ScaledMessageHash": [
    "0x47a3b4cd9bdd27685842848cf03d24e9a36baaa740ea0dd6f0460dc5f562ea2e",
    "0x0fc622367e2efe2fb12a60750f6b1e08fb352a766209158cf74410923cbaf628"
  ]
}
$ _build/default/src/app/cli/src/mina.exe advanced vrf generate-witness --delegator-index 0 --epoch-seed 2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA --global-slot 0 --privkey-path ~/.mina-config/wallets/store/B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh --delegated-stake 66000 --total-stake 66000 --generate-outputs true
Enter password:
Again to confirm:
{
  "message": {
    "globalSlot": "0",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  },
  "publicKey": "B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh",
  "c":
    "5589148566318731109878046041354274823231580307675604710374617202622632200642",
  "s":
    "8716271501507945010846517099629752945113631716814772175354793003714757922493",
  "ScaledMessageHash": [
    "0x47a3b4cd9bdd27685842848cf03d24e9a36baaa740ea0dd6f0460dc5f562ea2e",
    "0x0fc622367e2efe2fb12a60750f6b1e08fb352a766209158cf74410923cbaf628"
  ],
  "vrfThreshold": { "delegatedStake": "0.000066", "totalStake": "0.000066" },
  "vrfOutput": "EiRsVz2J6uM97Rj3Q8V6VNC2wRNmgDdce9z5BBUxjDDB1CXgTVxYE",
  "vrfOutputFractional": 0.2946216255238925,
  "thresholdMet": true
}
$ _build/default/src/app/cli/src/mina.exe advanced vrf generate-witness --delegator-index 0 --epoch-seed 2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA --global-slot 1 --privkey-path ~/.mina-config/wallets/store/B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh --delegated-stake 66000 --total-stake 66000 --gen
erate-outputs true
Enter password:
Again to confirm:
{
  "message": {
    "globalSlot": "1",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  },
  "publicKey": "B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh",
  "c":
    "9845720674638630122236009887302157574041010663771399677378236341600413530962",
  "s":
    "27390803442532152847809531698062932407873536147069373400974975216628288796150",
  "ScaledMessageHash": [
    "0x5ca9300446524ebc90e68b94ba0d875ef9002c0e819ebfe4900b3d744dee933e",
    "0xdcd9bf7a0d3899f3823668ed7a0f0ce93269f0baa0a0c0e483e8ca3c88e6640a"
  ],
  "vrfThreshold": { "delegatedStake": "0.000066", "totalStake": "0.000066" },
  "vrfOutput": "EiRtKcFdDpbz3Y86aFgjo4MMswQAsKLG1yQ9WMhigEYYZ498sN6fj",
  "vrfOutputFractional": 0.8041703997154197,
  "thresholdMet": false
}
$ _build/default/src/app/cli/src/mina.exe advanced vrf batch-generate-witness --privkey-path ~/.mina-config/wallets/store/B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh
Enter password:
Again to confirm:
{
    "globalSlot": "0",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  }
{
  "message": {
    "globalSlot": "0",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  },
  "publicKey": "B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh",
  "c":
    "24830300983323534205319769636733187797591537605083952937894345305780338349250",
  "s":
    "28232781266090283977328918442495885508107183663272421224647693174376986444951",
  "ScaledMessageHash": [
    "0x47a3b4cd9bdd27685842848cf03d24e9a36baaa740ea0dd6f0460dc5f562ea2e",
    "0x0fc622367e2efe2fb12a60750f6b1e08fb352a766209158cf74410923cbaf628"
  ]
}
{
    "globalSlot": "1",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
}
{
  "message": {
    "globalSlot": "1",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  },
  "publicKey": "B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh",
  "c":
    "21697049471984844987025477654590474208429717281912760602559000419271587574642",
  "s":
    "23429019983955541305864647597607685559697986199964846835173118443822846450182",
  "ScaledMessageHash": [
    "0x5ca9300446524ebc90e68b94ba0d875ef9002c0e819ebfe4900b3d744dee933e",
    "0xdcd9bf7a0d3899f3823668ed7a0f0ce93269f0baa0a0c0e483e8ca3c88e6640a"
  ]
}
$ _build/default/src/app/cli/src/mina.exe advanced vrf batch-check-witness
{
  "message": {
    "globalSlot": "0",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  },
  "publicKey": "B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh",
  "c":
    "24830300983323534205319769636733187797591537605083952937894345305780338349250",
  "s":
    "28232781266090283977328918442495885508107183663272421224647693174376986444951",
  "ScaledMessageHash": [
    "0x47a3b4cd9bdd27685842848cf03d24e9a36baaa740ea0dd6f0460dc5f562ea2e",
    "0x0fc622367e2efe2fb12a60750f6b1e08fb352a766209158cf74410923cbaf628"
  ]
}
{
  "message": {
    "globalSlot": "0",
    "epochSeed": "2va9BGv9JrLTtrzZttiEMDYw1Zj6a6EHzXjmP9evHDTG3oEquURA",
    "delegatorIndex": 0
  },
  "publicKey": "B62qiqVsJmi5St5JwjsveJa9e3EUsqdcS5ZQxSszurXGs7Nr4Na1VDh",
  "c":
    "24830300983323534205319769636733187797591537605083952937894345305780338349250",
  "s":
    "28232781266090283977328918442495885508107183663272421224647693174376986444951",
  "ScaledMessageHash": [
    "0x47a3b4cd9bdd27685842848cf03d24e9a36baaa740ea0dd6f0460dc5f562ea2e",
    "0x0fc622367e2efe2fb12a60750f6b1e08fb352a766209158cf74410923cbaf628"
  ],
  "vrfOutput": "EiRsVz2J6uM97Rj3Q8V6VNC2wRNmgDdce9z5BBUxjDDB1CXgTVxYE",
  "vrfOutputFractional": 0.2946216255238925
}
```

Checklist:

- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [X] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [X] Does this close issues? List them:

This is the bulk of the work for supporting #8677

### CODE REVIEW

File: src/lib/mina_graphql/mina_graphql.ml [hunk 1]
```diff
@@ -1100,7 +1130,11 @@ module Types = struct
                    | Token_owned _ ->
                        false
                    | Not_owned {account_disabled} ->
-                       account_disabled ) ] ))
+                       account_disabled )
+             ; field "index" ~typ:int
+                 ~doc:"The index of this account in the ledger"
```
Comment (bkase) [inline src/lib/mina_graphql/mina_graphql.ml:1135]: Can you add to the doc the cases when it is None? (in this case when the best_tip_ledger isn't Active, but also try an explain when this happens in practice from an operator's perspective)

File: src/lib/mina_graphql/mina_graphql.ml [hunk 2]
```diff
@@ -1961,6 +1995,64 @@ module Types = struct
...
 
+    let vrf_message =
+      obj "VrfMessageInput" ~doc:"The inputs to a vrf evaluation"
+        ~coerce:(fun global_slot epoch_seed delegator_index ->
+          { Consensus_vrf.Layout.Message.global_slot
+          ; epoch_seed= Mina_base.Epoch_seed.of_base58_check_exn epoch_seed
+          ; delegator_index } )
+        ~fields:
+          [ arg "globalSlot" ~typ:(non_null uint32_arg)
+          ; arg "epochSeed" ~typ:(non_null string)
```
Comment (bkase) [inline src/lib/mina_graphql/mina_graphql.ml:2006]: Can you add doc: "Formatted with base58check"

File: src/lib/cli_lib/commands.ml [hunk 3]
```diff
@@ -140,3 +140,199 @@ let validate_transaction =
...
+            | Error err ->
+                Format.eprintf "Error:@.%s@.@."
+                  (Yojson.Safe.pretty_to_string
+                     (Error_json.error_to_yojson err)) ;
+                `Repeat () )
+      in
+      exit 0 )
+
+  let command_group =
+    Command.group ~summary:"Commands for vrf evaluations"
```
Comment (bkase) [inline src/lib/cli_lib/commands.ml:334]: In the summary here (and on a relevant section of the GraphQL doc too) can you include more information on exactly how one could use these commands in practice -- I think these two scenarios cover all the bases (but up to you, if you want to choose different examples here) (1) figure out all of the slots that I will win for this current epoch or (2) for some epoch, ingest a 3rd party's batch witness and report which slots they have won.

We can then digest these docs into a page on the website too, but I think it would be really great if this lived in the help message and the GraphQL schema docs too

**Review by bkase — CHANGES_REQUESTED** [review]
Exciting!

**Comment by bkase** [comment]
Re the archive node change -- yes definitely, I made a new issue for this and we can discuss in prio. Luckily, we should be able to patch the existing data with a tweak to the replayer I think (cc @psteckler ).

**Review by bkase — APPROVED** [review]
Great

File: src/lib/vrf_lib/standalone.ml [hunk 4]
```diff
@@ -77,15 +77,15 @@ module Make
     end) (Message : sig
       open Impl
 
-      type t [@@deriving sexp]
```
Comment (imeckler) [inline src/lib/vrf_lib/standalone.ml:80]: Why this change out of curiosity?

**Review by imeckler — APPROVED** [review]

File: src/lib/vrf_lib/standalone.ml (hunk 4 above)
Comment (mrmr1993) [reply src/lib/vrf_lib/standalone.ml:80]: This was to make it match the types that we pass to the `integrated` version, so that I didn't have to recreate the module arguments with slightly different layouts for each.
